src/parseTools.js

```diff
@@ -1253,5 +1253,7 @@ function getUnsharedTextDecoderView(heap, start, end) {
 
   // Otherwise, generate a runtime type check: must do a .slice() if looking at a SAB,
   // or can use .subarray() otherwise.
-  return `${heap}.buffer instanceof SharedArrayBuffer ? ${shared} : ${unshared}`;
+  // In some cases, `instanceof SharedArrayBuffer` returns false even though buffer is an SAB.
+  // See: https://github.com/emscripten-core/emscripten/issues/15217
+  return `Object.prototype.toString.call(${heap}.buffer) === "[object SharedArrayBuffer]" ? ${shared} : ${unshared}`;
```

Comment:
Seems reasonable to me.  Thanks for linking the bug.   I wonder if we could add a test to test_browser.py?

Reply:
This reads madness to me.

The iframe gotcha with instanceof is something that I learned new now, and it makes sense, however that raises a few questions:

1. instanceof fails if `x instanceof y` is called with `x` and `y` that split iframe domains. In this case, why are `${heap}.buffer` and `SharedArrayBuffer` from two different iframe domains? This code should be all emitted inside the same JS domain, so the values should be the same. (obviously that is not the case since people are hitting this issue, but we should look at a more detailed repro first)
2. I am suspicious whether this call to `.toString()` would be standardized to require returning a string in that format? Is that confirmed? Or does it just happen to return the same value on current environments we care about?
3. We have 75 other uses of `instanceof` under JS files in `src/`. Which uses of `instanceof` should we ban there too?
4. The reason for the earlier **fast** type check was that opportunistically if the conditions match, we can avoid duplicating the string. However, this new check looks really slow, as it will need to do a string compare operation. Most strings that are converted are typically relatively short, so it is likely that just unconditionally performing the copy-to-unshared view operation will be much faster.

I think before we land a fix like this, we should take a look at point 1 first: why is it possible that we get values from splitting domains here when people are building into iframes? Is such data flow usage sensible, or is that something that should be avoided?

And if it is sensible, then look at the point 4 second: should we just unconditionally `return unshared;` there for better performance - the opportunistic "avoid-a-copy" optimization should not be turned into a slowdown.

And also if this kind of change is sensible, we do then have those 75 other call sites that can be suspect as well - can data in them split iframe domains?

I think we really should place a test for this, one that we accept that the use case is sound. If there are other places that are needed to be fixed for in-iframe execution, then we might even look if that would warrant a compile time flag `-sENVIRONMENT=web,worker,...,iframe`?